File: tasks/task2_toc_generator/cpp/solve.cpp
// Markdown TOC generator (C++17)
// Extracts headings from a file and outputs JSON structure.

#include <algorithm>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <map>
#include <optional>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

static std::string trim(const std::string& s) {
    size_t a = s.find_first_not_of(" \t\r\n");
    if (a == std::string::npos) return "";
    size_t b = s.find_last_not_of(" \t\r\n");
    return s.substr(a, b - a + 1);
}

// Extract heading level (number of # at start) and title
static std::optional<std::pair<int, std::string>> try_parse_heading(const std::string& line, int max_depth) {
    int lvl = 0;
    for (char c : line) {
        if (c == '#') {
            lvl++;
            if (lvl > max_depth) return std::nullopt;
        } else if (c == ' ' || c == '\t') {
            break;
        } else {
            return std::nullopt; // not a heading
        }
    }
    if (lvl == 0) return std::nullopt;
    // Find title after #... followed by space
    size_t pos = lvl;
    while (pos < line.size() && (line[pos] == ' ' || line[pos] == '\t')) pos++;
    if (pos >= line.size()) return std::nullopt;
    return std::make_pair(lvl, trim(line.substr(pos)));
}

struct Node {
    std::string title;
    int level = 0;
    std::vector<Node> children;
};

static std::string escape_json_str(const std::string& s) {
    std::string out;
    for (char c : s) {
        if (c == '"') out += "\\\"";
        else if (c == '\\') out += "\\\\";
        else out += c;
    }
    return out;
}

static std::string node_to_json(const Node& n, int indent) {
    std::string pad(indent, ' ');
    std::string pad2(indent + 2, ' ');
    std::string result = pad + "{\n";
    result += pad2 + "\"title\": \"" + escape_json_str(n.title) + "\",\n";
    result += pad2 + "\"level\": " + std::to_string(n.level) + ",\n";
    if (n.children.empty()) {
        result += pad2 + "\"children\": []\n";
    } else {
        result += pad2 + "\"children\": [\n";
        for (size_t i = 0; i < n.children.size(); i++) {
            result += node_to_json(n.children[i], indent + 4);
            if (i + 1 < n.children.size()) result += ",";
            result += "\n";
        }
        result += pad2 + "]\n";
    }
    result += pad + "}";
    return result;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: solve <file.md> [--output <path>] [--max-depth <n>]\n";
        return 1;
    }

    std::string filepath = argv[1];
    std::string output_path;
    int max_depth = 999;

    for (int i = 2; i < argc; i++) {
        std::string arg(argv[i]);
        if (arg == "--output" && i + 1 < argc) {
            output_path = argv[i + 1];
            i++;
        } else if (arg == "--max-depth" && i + 1 < argc) {
            max_depth = std::stoi(argv[i + 1]);
            i++;
        }
    }

    std::ifstream infile(filepath);
    if (!infile.is_open()) {
        std::cerr << "Cannot open " << filepath << "\n";
        return 1;
    }

    // Parse headings
    std::vector<std::pair<int, std::string>> headings;
    std::string line;
    while (std::getline(infile, line)) {
        auto h = try_parse_heading(trim(line), max_depth);
        if (h) headings.push_back(*h);
    }

    // Build nested TOC: skip H1 as item, H2+ as top-level
    std::vector<Node> root;
    std::vector<std::pair<int, Node*>> stack;

    for (auto& [lvl, title] : headings) {
        Node node{title, lvl, {}};

        if (lvl == 1) {
            // H1 is the title, skip as an item
            continue;
        }

        // Pop stack until we find a parent with lower level
        while (!stack.empty() && stack.back().first >= lvl)
            stack.pop_back();

        if (!stack.empty())
            stack.back().second->children.push_back(node);
        else
            root.push_back(node);

        stack.emplace_back(lvl, &root.back());
    }

    // Output
    std::string json_out = "[\n";
    for (size_t i = 0; i < root.size(); i++) {
        json_out += node_to_json(root[i], 2);
        if (i + 1 < root.size()) json_out += ",";
        json_out += "\n";
    }
    json_out += "]\n";

    if (!output_path.empty()) {
        std::ofstream out(output_path);
        out << json_out;
    } else {
        std::cout << json_out;
    }

    return 0;
}

File: tasks/task2_toc_generator/python/solve.py
#!/usr/bin/env python3
"""Markdown TOC generator.
Extracts headings from a markdown file and produces a structured TOC.
"""
import argparse
import json
import os
import re
import sys


def extract_headings(filepath: str, max_depth: int = 999):
    """Extract headings up to *max_depth* from a markdown file."""
    headings = []
    with open(filepath, "r") as f:
        for line in f:
            m = re.match(r'^(#{1,' + str(max_depth) + r'})\s+(.+)$', line)
            if m:
                level = len(m.group(1))
                title = m.group(2).strip()
                headings.append((level, title))
    return headings


def build_toc(headings, max_depth=999):
    """Build flat list of TOC items with children for nested structure.

    H2+ headings become top-level items. H1 is treated as title only
    and does not appear as an item (H2s are the first items).
    H3+ become children of their nearest H1/H2 ancestor.
    """
    # Separate H1 as title (not an item)
    root = []
    h1_title = None
    stack = []  # (level, node)

    for level, title in headings:
        if level > max_depth:
            continue

        node = {"title": title, "level": level, "children": []}

        if level == 1:
            h1_title = title
            continue  # skip H1 as an item

        # Pop stack until we find a parent with lower level
        while stack and stack[-1][0] >= level:
            stack.pop()

        if stack:
            stack[-1][1]["children"].append(node)
        else:
            root.append(node)

        stack.append((level, node))

    return root


def toc_to_html(toc, min_level=1):
    """Convert TOC list to HTML string."""
    parts = []
    for item in toc:
        level = item["level"]
        indent = "  " * (level - min_level)
        anchor = item["title"].lower().replace(" ", "-")
        parts.append(f'{indent}<li><a href="#{anchor}">{item["title"]}</a>')
        if item["children"]:
            parts.append(f'{indent}<ul>')
            parts.append(toc_to_html(item["children"], level + 1))
            parts.append(f'{indent}</ul>')
        parts.append(f'{indent}</li>')
    return "\n".join(parts)


def main():
    parser = argparse.ArgumentParser(description="Generate a TOC from a markdown file.")
    parser.add_argument("file", help="Path to markdown file")
    parser.add_argument("--output", "-o", help="Output file path")
    parser.add_argument("--max-depth", "-d", type=int, default=999, help="Max heading depth")
    parser.add_argument("--format", "-f", choices=["json", "html"], default="json", help="Output format")
    args = parser.parse_args()

    if not os.path.isfile(args.file):
        print(f"Error: {args.file} is not a file", file=sys.stderr)
        sys.exit(1)

    headings = extract_headings(args.file, args.max_depth)
    toc = build_toc(headings, args.max_depth)

    if args.format == "html":
        output = "<!DOCTYPE html>\n<html>\n<head><title>Table of Contents</title></head>\n<body>\n<ul>\n"
        output += toc_to_html(toc)
        output += "\n</ul>\n</body>\n</html>\n"
    else:
        output = json.dumps(toc, indent=2)

    if args.output:
        with open(args.output, "w") as f:
            f.write(output)
    else:
        print(output)


if __name__ == "__main__":
    main()
